### PR TITLE
Fix line numbers in macros

### LINKED ISSUE #1174: Errors in a @constraints block reported with incorrect line numbers
If you make an error in a `@constraints` block the line number reported to cause the error is the line of the `@constraints` statement. For example:
```julia
m = Model(solver=ClpSolver())

@variables m begin
    0 <= x <= 4
    0 <= y <= 30
end

@constraints m begin        # <- error reported on this line
    region1, 1x + 5y <= 3.0
    region2, x - undeclared_variable <= 2.0
end

@objective(m, Max, 5x + 3y)

print(m)
status = solve(m)
```

In this example the error is `ERROR: LoadError: UndefVarError: undeclared_variable not defined`, so you would easily find the problem anyway. But I often get more generic errors which don't have an immediately obvious cause, and if you have 20+ constraints then it becomes extremely difficult to identify the problem.

This greatly limits the usability of the `@constraints` syntax for large models - the precise situation in which it is intended to be convenient.

### PR DESCRIPTION
**Pull request by odow**
Closes #1174

### CODE REVIEW

**Comment by blegat** [comment]
So what does `__souce__` do?

**Comment by odow** [comment]
`__source__` is a hidden `LineNumberNode` that contains the file and line of the caller.

## If the error is at runtime

```julia
using JuMP
model = Model()
@variable(model, x)
@constraints(model, begin
    x <= 1
    y <= 2
end)
```

### Old stack trace
```Julia
ERROR: LoadError: UndefVarError: y not defined
Stacktrace:
 [1] top-level scope at /Users/oscar/.julia/packages/MutableArithmetics/NuiNA/src/rewrite.jl:227
 [2] top-level scope at /Users/oscar/.julia/packages/JuMP/YXK4e/src/macros.jl:416
 [3] include at ./boot.jl:328 [inlined]
 [4] include_relative(::Module, ::String) at ./loading.jl:1105
 [5] include(::Module, ::String) at ./Base.jl:31
 [6] include(::String) at ./client.jl:424
 [7] top-level scope at REPL[2]:1
in expression starting at /Users/oscar/.julia/dev/JuMP/err.jl:4
```

### New stack trace

Note the new frame **`[3] top-level scope at /Users/oscar/.julia/dev/JuMP/err.jl:6`**
```Julia
ERROR: LoadError: UndefVarError: y not defined
Stacktrace:
 [1] top-level scope at /Users/oscar/.julia/packages/MutableArithmetics/NuiNA/src/rewrite.jl:227
 [2] top-level scope at /Users/oscar/.julia/dev/JuMP/src/macros.jl:421
 [3] top-level scope at /Users/oscar/.julia/dev/JuMP/err.jl:6
 [4] include(::String) at ./client.jl:439
 [5] top-level scope at REPL[2]:1
in expression starting at /Users/oscar/.julia/dev/JuMP/err.jl:4
```

## If the error is at compile time

```julia
using JuMP
model = Model()
@variable(model, x)
@constraints(model, begin
    x <= 1
    y
end)
```

### Old

```Julia
ERROR: LoadError: LoadError: In `@constraint(model, y)`: Incomplete constraint specification y. Are you missing a comparison (<=, >=, or ==)?
Stacktrace:
 [1] error(::String, ::String) at ./error.jl:42
 [2] _macro_error(::Symbol, ::Array{Symbol,1}, ::String) at /Users/oscar/.julia/packages/JuMP/YXK4e/src/macros.jl:894
 [3] (::JuMP.var"#_error#70"{Symbol})(::String) at /Users/oscar/.julia/packages/JuMP/YXK4e/src/macros.jl:366
 [4] _constraint_macro(::Tuple{Symbol,Symbol}, ::Symbol, ::typeof(parse_constraint_expr)) at /Users/oscar/.julia/packages/JuMP/YXK4e/src/macros.jl:397
 [5] @constraint(::LineNumberNode, ::Module, ::Vararg{Any,N} where N) at /Users/oscar/.julia/packages/JuMP/YXK4e/src/macros.jl:490
 [6] include at ./boot.jl:328 [inlined]
 [7] include_relative(::Module, ::String) at ./loading.jl:1105
 [8] include(::Module, ::String) at ./Base.jl:31
 [9] include(::String) at ./client.jl:424
 [10] top-level scope at REPL[2]:1
in expression starting at /Users/oscar/.julia/dev/JuMP/err.jl:6
in expression starting at /Users/oscar/.julia/dev/JuMP/err.jl:4
```

### New

We now print the line **`At /Users/oscar/.julia/dev/JuMP/err.jl:6:`** as part of the error message.

```Julia
ERROR: LoadError: LoadError: At /Users/oscar/.julia/dev/JuMP/err.jl:6: `@constraint(model, y)`: Incomplete constraint specification y. Are you missing a comparison (<=, >=, or ==)?
Stacktrace:
 [1] error(::String, ::String) at ./error.jl:42
 [2] _macro_error(::Symbol, ::Array{Symbol,1}, ::LineNumberNode, ::String) at /Users/oscar/.julia/dev/JuMP/src/macros.jl:903
 [3] (::JuMP.var"#_error#70"{Symbol,LineNumberNode})(::String) at /Users/oscar/.julia/dev/JuMP/src/macros.jl:371
 [4] _constraint_macro(::Tuple{Symbol,Symbol}, ::Symbol, ::typeof(parse_constraint_expr), ::LineNumberNode) at /Users/oscar/.julia/dev/JuMP/src/macros.jl:402
 [5] @constraint(::LineNumberNode, ::Module, ::Vararg{Any,N} where N) at /Users/oscar/.julia/dev/JuMP/src/macros.jl:495
 [6] include(::String) at ./client.jl:439
 [7] top-level scope at REPL[2]:1
in expression starting at /Users/oscar/.julia/dev/JuMP/err.jl:6
in expression starting at /Users/oscar/.julia/dev/JuMP/err.jl:4
```

File: src/macros.jl [hunk 1]
```diff
@@ -38,12 +38,14 @@ end
...
-function _assert_valid_model(m, macrocode)
+function _assert_valid_model(m, macrocode, source)
     # assumes m is already escaped
-    quote
-        _valid_model($m, $(quot(m.args[1])))
-        $macrocode
-    end
+    return Expr(
+        :block,
+        source,
```
Comment (mlubin) [inline src/macros.jl:45]: This deserves a comment explaining what the block does.

File: src/macros.jl [hunk 2]
```diff
@@ -365,8 +367,8 @@ determine the container type) are added to the `build_constraint` call. The
 returned value of this call is passed to `add_constraint` which returns a
 constraint reference.
 """
-function _constraint_macro(args, macro_name::Symbol, parsefun::Function)
-    _error(str...) = _macro_error(macro_name, args, str...)
+function _constraint_macro(args, macro_name::Symbol, parsefun::Function, source)
```
Comment (mlubin) [inline src/macros.jl:370]: The arguments no longer match the docstring.

File: src/macros.jl [hunk 3]
```diff
@@ -38,12 +38,14 @@ end
 _valid_model(m::AbstractModel, name) = nothing
 _valid_model(m, name) = error("Expected $name to be a JuMP model, but it has type ", typeof(m))
 
-function _assert_valid_model(m, macrocode)
+function _assert_valid_model(m, macrocode, source)
```
Comment (mlubin) [inline src/macros.jl:41]: This function may need a new name. `_finalize_macro`?

Comment (mlubin) [inline src/macros.jl:41]: `source_line::LineNumberNode` would be a bit more readable.

**Comment by odow** [comment]
I just realized I forgot to fix the line numbers in the macros inside `JuMP.Containers`, but I can do that in a separate PR.

**Pull request merged**